File: examples/python/sphincs-wallet/SPHINCSWallet.runar.py
from runar import SmartContract, ByteString, public, assert_, verify_slh_dsa_sha2_128s

class SPHINCSWallet(SmartContract):
    pubkey: ByteString

    def __init__(self, pubkey: ByteString):
        super().__init__(pubkey)
        self.pubkey = pubkey

    @public
    def spend(self, msg: ByteString, sig: ByteString):
        assert_(verify_slh_dsa_sha2_128s(msg, sig, self.pubkey))

File: examples/zig/sphincs-wallet/SPHINCSWallet.runar.zig
const runar = @import("runar");

pub const SPHINCSWallet = struct {
    pub const Contract = runar.SmartContract;

    ecdsaPubKeyHash: runar.Addr,
    slhdsaPubKeyHash: runar.ByteString,

    pub fn init(ecdsaPubKeyHash: runar.Addr, slhdsaPubKeyHash: runar.ByteString) SPHINCSWallet {
        return .{
            .ecdsaPubKeyHash = ecdsaPubKeyHash,
            .slhdsaPubKeyHash = slhdsaPubKeyHash,
        };
    }

    pub fn spend(
        self: *const SPHINCSWallet,
        slhdsaSig: runar.ByteString,
        slhdsaPubKey: runar.ByteString,
        sig: runar.Sig,
        pubKey: runar.PubKey,
    ) void {
        runar.assert(runar.bytesEq(runar.hash160(pubKey), self.ecdsaPubKeyHash));
        runar.assert(runar.checkSig(sig, pubKey));
        runar.assert(runar.bytesEq(runar.hash160(slhdsaPubKey), self.slhdsaPubKeyHash));
        runar.assert(runar.verifySLHDSA_SHA2_128s(sig, slhdsaSig, slhdsaPubKey));
    }
};
